src/server/pkg/storage/chunk/reader.go

```diff
@@ -4,202 +4,330 @@ import (
+	// Use seed chunk if possible.
+	if dr.seed != nil && dr.dataRef.ChunkInfo.Chunk.Hash == dr.seed.dataRef.ChunkInfo.Chunk.Hash {
+		if err := dr.seed.getChunk(); err != nil {
 			return err
 		}
-		totalLeft -= n
+		dr.chunk = dr.seed.chunk
 		return nil
 	}
-	c := &Copy{
-		before: &bytes.Buffer{},
-		after:  &bytes.Buffer{},
+	// Get chunk from object storage.
+	objR, err := dr.objC.Reader(dr.ctx, path.Join(prefix, dr.dataRef.ChunkInfo.Chunk.Hash), 0, 0)
+	if err != nil {
+		return err
 	}
-	// Copy the first WindowSize bytes raw to be sure that
-	// the chunks we will copy will exist in the writer that
-	// this data is being copied to.
-	if err := rawCopy(c.before, mathutil.MinInt64(WindowSize, totalLeft)); err != nil {
-		return nil, err
+	defer objR.Close()
+	gzipR, err := gzip.NewReader(objR)
+	if err != nil {
+		return err
 	}
-	// Copy the bytes left in the current chunk (if any)
-	if r.r.Len() > 0 {
-		if err := rawCopy(c.before, mathutil.MinInt64(int64(r.r.Len()), totalLeft)); err != nil {
-			return nil, err
-		}
+	defer gzipR.Close()
+	buf := &bytes.Buffer{}
+	if _, err := io.Copy(buf, gzipR); err != nil {
+		return err
 	}
-	// Copy the in between chunk references.
-	// (bryce) is there an edge case with a size zero chunk?
-	for len(r.dataRefs) > 0 && r.dataRefs[0].Hash == "" && totalLeft >= r.dataRefs[0].SizeBytes {
-		r.executeOnSplitFunc()
-		c.chunkRefs = append(c.chunkRefs, r.dataRefs[0])
-		totalLeft -= r.dataRefs[0].SizeBytes
-		r.len -= r.dataRefs[0].SizeBytes
-		r.dataRefs = r.dataRefs[1:]
+	dr.chunk = buf.Bytes()
+	return nil
+}
+
+// BeforeBound checks if the passed in string is before the string bound (exclusive).
+// The string bound is optional, so if no string bound is passed then it returns true.
+func BeforeBound(str string, strBound ...string) bool {
+	return len(strBound) == 0 || str < strBound[0]
+}
+
+// Get writes the data referenced by the data reference.
+func (dr *DataReader) Get(w io.Writer) error {
+	if err := dr.getChunk(); err != nil {
+		return err
 	}
-	// Copy the rest of the bytes raw.
-	if err := rawCopy(c.after, totalLeft); err != nil {
-		return nil, err
+	data := dr.chunk[dr.dataRef.OffsetBytes : dr.dataRef.OffsetBytes+dr.dataRef.SizeBytes]
+	if _, err := w.Write(data); err != nil {
+		return err
 	}
-	return c, nil
+	return nil
 }
 
-// Close closes the reader.
-// Currently a no-op, but will be used when streaming is implemented.
-func (r *Reader) Close() error {
-	return nil
+// LimitReader creates a new data reader that reads a subset of the remaining tags in the current data reader.
+// This tag subset is determined by the optional parameter tagUpperBound which specifies the upper bound (exclusive) of the new data reader.
+// LimitReader will progress the current data reader past the tags in the tag subset.
+// Data in the tag subset is fetched lazily by the new data reader.
+func (dr *DataReader) LimitReader(tagUpperBound ...string) *DataReader {
```

Comment:
How would you feel about calling this `BoundReader` because `LimitReader` already exists in the standard library, and has connotations of limiting based on size, whereas this is based on a tag bound, so I think it's a slightly clearer name.

Reply:
Makes sense, done.